src/magic_folder/test/test_magic_folder.py

```diff
@@ -2452,3 +2471,108 @@ def setUp(self):
         d = super(RealTestAliceBob, self).setUp()
         self.inotify = get_inotify_module()
         return d
+
+class SnapshotTest(AsyncTestCase):
```

Comment:
I'd like to try to find a better testing strategy than "launch Tahoe-LAFS, throw some HTTP at it".  These tests are going to be slow, failure prone, and a pain in the ass to debug when they fail.  All of these things will also deny us the ability to have really in-depth testing of all of the edge-cases for the snapshot-handling tests.

Reply:
treq has some built-in testing support, like `treq.testing.RequestTraversalAgent` that could probably be leveraged to provide a fake Tahoe API -- that itself is probably a whole PR, maybe, but could be a useful direction to make the tests more immediate and less directly tied to "a running tahoe"...

Reply:
I will explore the `RequestTraversalAgent` option. Yes, launching Tahoe-LAFS node definitely looks ugly and it was very hard to debug even simple type errors that I was making while doing the development. An option is the "sans-io" type of approach where we generate Tahoe requests as messages?

Reply:
The "ideal" situation here would be if we had a "verified fake" Tahoe API endpoint .. that is, the "root" that we can give to `RequestTraversalAgent`. Then we'd be doing in-memory requests to a fake version of a Tahoe node (for any API requests).

The "verified" part I guess would come from integration-style tests (that actually start a Tahoe and confirm that our "fake" does the same thing as the real one). I am thinking of this article when talking about it: https://pythonspeed.com/articles/verified-fakes/

However, we do not have that. This would be a whole PR (or more!) by itself -- maybe this is something I could help with. At first, we could cover *just* enough to get some tests for snapshots. Meantime, I think the best approach for this PR is to get as much coverage from unit-style tests on the helpers and so on as we can.

...it is maybe even premature to write such a test? That is, if we can't write an integration test for snapshots yet (which I think is true?) then can we write this "verified fake" style of test yet?

Reply:
maybe moot though since there isn't such a python API ... yet?

Reply:
Yes, I think such an API would be great to have. I'd previously proposed this as a way to "greenfield" a namespace without `allmydata` too (i.e. at first the only public stuff in such an API would be the minimum needed to do public Tahoe stuff) .. but, yeah, there isn't such an API.

On the flip side, since `tahoe` can only be interacted with via the command-line and WebUI there is no need for it to be a Python program at all ;)